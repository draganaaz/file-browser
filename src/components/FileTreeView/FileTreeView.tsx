import React, { useState, useRef, useCallback } from 'react';
import {
  FILE_TYPE,
  MENU_ITEMS,
  VALID_EXTENSIONS,
} from '../../constants/fileTree';
import { FileNode } from '../../types/FileNode';
import { isValidFileName } from '../../utils/validations';
import Modal from '../Modal/Modal';
import { useOutsideClick } from '../../hooks/useOutsideClick';
import { findNodeById } from '../../utils/treeService';

interface FileTreeViewProps {
  data: FileNode[];
  onAdd: (parentId: string, name: string, type: FILE_TYPE) => void;
  onDelete: (nodeId: string) => void;
  onSelect: (node: FileNode, path: string[]) => void;
  onRename: (nodeId: string, newName: string) => void;
  selectedNode: FileNode | null;
}

const FileTreeView: React.FC<FileTreeViewProps> = ({
  data,
  onAdd,
  onDelete,
  onRename,
  onSelect,
  selectedNode,
}) => {
  const [menuNodeId, setMenuNodeId] = useState<string | null>(null);
  const [submenuNodeId, setSubmenuNodeId] = useState<string | null>(null);
  const [activeParentId, setActiveParentId] = useState<string | null>(null);
  const [newItemName, setNewItemName] = useState('');
  const [errorMessage, setErrorMessage] = useState('');
  const [fileExtension, setFileExtension] = useState<string | null>(null);
  const [itemType, setItemType] = useState<FILE_TYPE | null>(null);
  const [isRenaming, setIsRenaming] = useState(false);
  const [showDetails, setShowDetails] = useState(false);

  const menuRef = useRef<HTMLDivElement>(null);

  const closeMenus = useCallback(() => {
    setMenuNodeId(null);
    setSubmenuNodeId(null);
    setErrorMessage('');
    setIsRenaming(false);
    setShowDetails(false);
  }, []);

  useOutsideClick(menuRef, () => closeMenus());

  const handleAdd = (parentId: string, type: FILE_TYPE, extension?: string) => {
    setActiveParentId(parentId);
    setItemType(type);
    extension && setFileExtension(extension);
    closeMenus();
  };

  const handleConfirmAdd = (parentId: string) => {
    let name = newItemName.trim();

    // When no name provided, discard the input
    if (!name) {
      setActiveParentId(null);
      setNewItemName('');
      setErrorMessage('');
      return;
    }

    if (itemType === FILE_TYPE.FILE) {
      const nameParts = name.split('.');
      const nameExtension =
        nameParts.length > 1 ? nameParts.pop() : fileExtension;

      if (
        nameExtension &&
        !Object.values(VALID_EXTENSIONS).includes(nameExtension)
      ) {
        setErrorMessage(
          'Invalid file extension. Please use a valid extension.'
        );
        return;
      }
      name = nameParts.join('.') + '.' + nameExtension;
    }

    if (!isValidFileName(name)) {
      setErrorMessage('Invalid name. Please avoid special characters.');
      return;
    }

    const parentFolder = data.find((node) => node.id === parentId);
    if (parentFolder?.children?.some((child) => child.name === name)) {
      setErrorMessage(
        'This item already exists at this location. Please choose a different name.'
      );
      return;
    }

    onAdd(parentId, name, itemType as FILE_TYPE);
    setActiveParentId(null);
    setNewItemName('');
    setErrorMessage('');
  };

  const handleRename = (nodeId: string, newName: string) => {
    // Discard any ongoing addition before proceeding to renaming
    setActiveParentId(null);

    const node = findNodeById(data, nodeId);

    if (!node) {
      setErrorMessage('File/Folder not found.');
      return;
    }

    const nameParts = newName.trim().split('.');
    const newBaseName = nameParts.slice(0, -1).join('.') || nameParts[0];
    const newExtension = nameParts.length > 1 ? nameParts.pop() : null;

    if (node.type === FILE_TYPE.FILE) {
      const originalExtension = node.name.split('.').pop();

      // If no extension is provided, use the original one
      const finalExtension = newExtension || originalExtension;

      // Validate the new extension
      if (
        finalExtension &&
        !Object.values(VALID_EXTENSIONS).includes(finalExtension)
      ) {
        setErrorMessage(
          `Invalid file extension. You must use a valid extension.`
        );
        return;
      }

      newName = newBaseName + (finalExtension ? `.${finalExtension}` : '');
    } else {
      newName = newBaseName; // For folders, avoid adding any extensions.
    }

    // Validate the full name
    if (!isValidFileName(newName)) {
      setErrorMessage('Invalid name. Please avoid special characters.');
      return;
    }

    onRename(nodeId, newName);
    setIsRenaming(false);
    setErrorMessage('');
  };

  const handleModalClose = () => {
    setShowDetails(false);
    setMenuNodeId(null);
  };

  const handleDetailsClick = (nodeId: string) => {
    const node = data.find((node) => node.id === nodeId);
    if (node) {
      setShowDetails(true);
      setMenuNodeId(null);
    }
  };

  const renderAddInputField = (node: FileNode) => (
    <div className="ml-6 mt-2">
      <input
        type="text"
        placeholder={`Enter ${itemType === FILE_TYPE.FOLDER ? 'folder' : 'file'} name`}
        value={newItemName}
        onChange={(e) => setNewItemName(e.target.value)}
        onKeyDown={(e) => {
          e.key === 'Enter' && handleConfirmAdd(node.id);
        }}
        onBlur={() => {
          handleConfirmAdd(node.id);
        }}
        autoFocus
        className="border rounded p-2 w-full focus:outline-none focus:ring-2 focus:ring-blue-500"
      />
      {errorMessage && (
        <p className="text-red-500 text-sm mt-1">{errorMessage}</p>
      )}
    </div>
  );

  const renderRenameInputField = (node: FileNode) => (
    <div className="ml-6 mt-2">
      <input
        type="text"
        defaultValue={node.name}
        onBlur={(e) => handleRename(node.id, e.target.value)}
        onKeyDown={(e) =>
          e.key === 'Enter' && handleRename(node.id, e.currentTarget.value)
        }
        autoFocus
        className="border rounded p-2 w-full focus:outline-none focus:ring-2 focus:ring-blue-500"
      />
      {errorMessage && (
        <p className="text-red-500 text-sm mt-1">{errorMessage}</p>
      )}
    </div>
  );

  const renderMenuItems = (nodeId: string) => (
    <div
      className="absolute right-[-160px] top-0 bg-white border rounded shadow-md p-1 w-40 z-30 menu-container"
      ref={menuRef}
    >
      {MENU_ITEMS.map((item) => (
        <button
          key={item.label}
          onClick={() => handleAdd(nodeId, item.type, item.extension)}
          className="w-full px-2 py-1 text-left hover:bg-gray-200"
          data-testid="menu-item"
        >
          {item.label}
        </button>
      ))}
    </div>
  );

  const renderActionsMenu = (nodeId: string, isFolder: boolean) => (
    <div className="absolute right-0 mt-1 bg-white border rounded shadow-md p-1 w-36 z-20 menu-container">
      {isFolder && (
        <button
          onClick={() =>
            setSubmenuNodeId(submenuNodeId === nodeId ? null : nodeId)
          }
          className="w-full px-2 py-1 hover:bg-gray-200 text-left"
          data-testid="add-button"
        >
          Add New...
        </button>
      )}
      <button
        onClick={() => {
          setIsRenaming(true);
          setMenuNodeId(null);
        }}
        className="w-full px-2 py-1 hover:bg-gray-200 text-left"
        data-testid="rename-button"
      >
        Rename
      </button>
      <button
        onClick={() => {
          onDelete(nodeId);
          closeMenus();
        }}
        className="w-full px-2 py-1 hover:bg-gray-200 text-left text-red-500"
        data-testid="delete-button"
      >
        Delete
      </button>
      <button
        onClick={() => handleDetailsClick(nodeId)}
        className="w-full px-2 py-1 hover:bg-gray-200 text-left"
        data-testid="details-button"
      >
        Details
      </button>
      {submenuNodeId === nodeId && renderMenuItems(nodeId)}
    </div>
  );

  const renderTree = (nodes: FileNode[], path: string[] = ['My Files']) =>
    nodes.map((node) => {
      const newPath = [...path, node.name];

      return (
        <li key={node.id} className="relative mb-2">
          <div
            className="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
            onClick={() => onSelect(node, newPath)}
          >
            {/* Renaming node */}
            {isRenaming && selectedNode?.id === node.id ? (
              renderRenameInputField(node)
            ) : (
              // Displaying regular node
              <span data-testid={`folder-name-${node.id}`}>
                {node.type === FILE_TYPE.FOLDER ? 'üìÅ' : 'üìÑ'} {node.name}
              </span>
            )}
            <div className="relative">
              <button
                onClick={(e) => {
                  e.stopPropagation();
                  setMenuNodeId(menuNodeId === node.id ? null : node.id);
                }}
                className="text-gray-600 hover:text-gray-800 px-2 py-0.5"
                data-testid={`menu-button-${node.id}`}
              >
                ‚ãÆ
              </button>
              {menuNodeId === node.id &&
                renderActionsMenu(node.id, node.type === FILE_TYPE.FOLDER)}
            </div>
          </div>
          {activeParentId === node.id && renderAddInputField(node)}

          {/* Render subtree */}
          {node.children && (
            <ul className="ml-4">{renderTree(node.children, newPath)}</ul>
          )}
        </li>
      );
    });

  return (
    <>
      <ul className="p-2">
        {data.length > 0 ? renderTree(data) : <p>No files available</p>}
      </ul>
      {showDetails && selectedNode && (
        <Modal
          show={true}
          onClose={handleModalClose}
          title={`Details for ${selectedNode.name}`}
        >
          <div className="text-sm text-gray-600">
            <p>Created At: {selectedNode.createdAt?.toLocaleString()}</p>
          </div>
        </Modal>
      )}
    </>
  );
};

export default FileTreeView;
