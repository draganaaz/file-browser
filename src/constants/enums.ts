export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}
