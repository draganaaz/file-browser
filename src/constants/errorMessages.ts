export const ERROR_MESSAGES = {
  INVALID_NAME_OR_EXTENSION: 'Invalid name or extension.',
  ITEM_EXISTS:
    'This item already exists at this location. Please choose a different name.',
  FILE_NOT_FOUND: 'File/Folder not found.',
};
