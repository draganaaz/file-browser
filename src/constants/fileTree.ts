export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}

export const VALID_EXTENSIONS = {
  TEXT: 'txt',
  JSON: 'json',
  IMAGE: 'png',
};

export const MENU_ITEMS = [
  { label: 'New Folder', type: FILE_TYPE.FOLDER },
  {
    label: 'New Text Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.TEXT,
  },
  {
    label: 'New JSON Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.JSON,
  },
  {
    label: 'New Image',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.IMAGE,
  },
];
