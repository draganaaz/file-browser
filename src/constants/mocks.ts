import { FileNode } from '../types/FileNode';
import { FILE_TYPE } from './fileTree';

export const initialData: FileNode[] = [
  {
    id: '1',
    name: 'public',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '1-1',
        name: 'exus-logo.png',
        type: FILE_TYPE.FILE,
        fileContent: '/exus.png',
      },
    ],
  },
  {
    id: '2',
    name: 'server',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '2-1',
        name: 'config.json',
        type: FILE_TYPE.FILE,
        fileContent: '{"key": "value"}',
      },
      {
        id: '2-2',
        name: 'notes.txt',
        type: FILE_TYPE.FILE,
        fileContent: 'This is a sample text file.',
      },
    ],
  },
  {
    id: '3',
    name: 'src',
    type: FILE_TYPE.FOLDER,
    children: [],
  },
];
