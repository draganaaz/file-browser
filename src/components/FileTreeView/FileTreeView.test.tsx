import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import FileTreeView from './FileTreeView';
import { FILE_TYPE } from '../../constants/fileTree';
import { initialData } from '../../constants/mocks';

const mockOnAdd = jest.fn();
const mockOnDelete = jest.fn();
const mockOnSelect = jest.fn();
const mockOnRename = jest.fn();
const mockSetFileTree = jest.fn();

const mockInitialData = initialData;

jest.mock('../../contexts/FileTreeContext', () => ({
  useFileTree: () => ({
    fileTree: mockInitialData,
    setFileTree: mockSetFileTree,
    onAdd: mockOnAdd,
    onDelete: mockOnDelete,
    onSelect: mockOnSelect,
    onRename: mockOnRename,
  }),
}));

describe('FileTreeView Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the file tree correctly', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={null}
      />
    );

    expect(screen.getByTestId('folder-name-1')).toHaveTextContent('public');
    expect(screen.getByTestId('folder-name-2')).toHaveTextContent('server');
    expect(screen.getByTestId('folder-name-3')).toHaveTextContent('src');
  });

  it('displays action buttons (Add, Delete, Rename) on hover', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={null}
      />
    );

    const folderItem = screen.getByTestId('folder-name-1');
    fireEvent.mouseEnter(folderItem);

    expect(screen.getByTestId('add-button')).toBeInTheDocument();
    expect(screen.getByTestId('delete-button')).toBeInTheDocument();
    expect(screen.getByTestId('rename-button')).toBeInTheDocument();
  });

  it('opens the add submenu when Add is clicked', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={null}
      />
    );

    const folderItem = screen.getByTestId('folder-name-1');
    fireEvent.mouseEnter(folderItem);

    const addButton = screen.getByTestId('add-button');
    fireEvent.click(addButton);

    expect(screen.getByText('New Folder')).toBeInTheDocument();
    expect(screen.getByText('New Text Document')).toBeInTheDocument();
    expect(screen.getByText('New JSON Document')).toBeInTheDocument();
    expect(screen.getByText('New Image')).toBeInTheDocument();
  });

  it('calls onAdd when a new folder is added', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={null}
      />
    );

    const folderItem = screen.getByTestId('folder-name-1');
    fireEvent.mouseEnter(folderItem);

    const addButton = screen.getByTestId('add-button');
    fireEvent.click(addButton);

    const folderOption = screen.getByText('New Folder');
    fireEvent.click(folderOption);

    const inputField = screen.getByPlaceholderText('Enter folder name');
    fireEvent.change(inputField, { target: { value: 'newFolder' } });
    fireEvent.keyDown(inputField, { key: 'Enter', code: 'Enter' });

    expect(mockOnAdd).toHaveBeenCalledWith('1', 'newFolder', FILE_TYPE.FOLDER);
  });

  it('calls onDelete when Delete is clicked', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={null}
      />
    );

    const folderItem = screen.getByTestId('folder-name-1');
    fireEvent.mouseEnter(folderItem);

    const deleteButton = screen.getByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalledWith('1');
  });

  it('calls onRename when renaming a folder', () => {
    render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={initialData[0]}
      />
    );

    const folderItem = screen.getByTestId('folder-name-1');
    fireEvent.mouseEnter(folderItem);

    const renameButton = screen.getByTestId('rename-button');
    fireEvent.click(renameButton);

    const inputField = screen.getByDisplayValue('public');
    fireEvent.change(inputField, { target: { value: 'newPublic' } });
    fireEvent.keyDown(inputField, { key: 'Enter', code: 'Enter' });

    expect(mockOnRename).toHaveBeenCalledWith('1', 'newPublic');
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <FileTreeView
        onAdd={mockOnAdd}
        onDelete={mockOnDelete}
        onSelect={mockOnSelect}
        onRename={mockOnRename}
        selectedNode={initialData[0]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
