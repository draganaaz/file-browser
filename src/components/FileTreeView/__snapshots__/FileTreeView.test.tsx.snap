// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileTreeView Component matches the snapshot 1`] = `
<DocumentFragment>
  <ul
    class="p-2"
  >
    <li
      class="relative mb-2"
    >
      <div
        class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
      >
        <span
          data-testid="folder-name-1"
        >
          📁 public
        </span>
      </div>
      <ul
        class="ml-4"
      >
        <li
          class="relative mb-2"
        >
          <div
            class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
          >
            <span
              data-testid="folder-name-1-1"
            >
              📄 exus-logo.png
            </span>
          </div>
        </li>
      </ul>
    </li>
    <li
      class="relative mb-2"
    >
      <div
        class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
      >
        <span
          data-testid="folder-name-2"
        >
          📁 server
        </span>
      </div>
      <ul
        class="ml-4"
      >
        <li
          class="relative mb-2"
        >
          <div
            class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
          >
            <span
              data-testid="folder-name-2-1"
            >
              📄 config.json
            </span>
          </div>
        </li>
        <li
          class="relative mb-2"
        >
          <div
            class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
          >
            <span
              data-testid="folder-name-2-2"
            >
              📄 notes.txt
            </span>
          </div>
        </li>
      </ul>
    </li>
    <li
      class="relative mb-2"
    >
      <div
        class="flex items-center justify-between p-2 rounded hover:bg-gray-200 cursor-pointer"
      >
        <span
          data-testid="folder-name-3"
        >
          📁 src
        </span>
      </div>
      <ul
        class="ml-4"
      />
    </li>
  </ul>
</DocumentFragment>
`;
